IZP2tester_zs2020.py

```diff
@@ -18,87 +19,87 @@
-        self.valgrind = valgrind
-        self.valgrind_stack = valgrind_stack
+        self.file_input = file_input
 
-    def run_test(self):
-        with tempfile.NamedTemporaryFile() as test_input_file, open(self.process_input, 'rb') as source_input_file:
-            test_input_file.write(source_input_file.read())
-            test_input_file.flush()
-            try:
-                program_with_args = [f"./{self.exe_path}"] + self.args + [test_input_file.name]
-                subprocess.run(program_with_args, timeout=1, check=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
-                test_input_file.seek(0)
-                self.actual_output = test_input_file.read().decode("utf-8")
-
-            except UnicodeDecodeError:
-                self.actual_output = f"BINARY_TRASH\n\n"
-            except TimeoutError:
-                self.actual_output = f"TIMED OUT\n\n"
-            except subprocess.CalledProcessError as err:
-                # -11 represents SEGFAULT: https://code-examples.net/en/q/11dd30f
-                self.actual_output = (f'SEGFAULT' if err.returncode == -11 else f'ERROR') + '\n'
-
-
-            if self.valgrind:
-                test_input_file.seek(0)
-                source_input_file.seek(0)
-                test_input_file.truncate()
-                test_input_file.write(source_input_file.read())
-                test_input_file.flush()
-
-
-                cmd_line = ['valgrind', '--log-fd=1', '-q', '--leak-check=full'] + \
-                    (['--max-stackframe=4040064'] if self.valgrind_stack else []) + \
-                    program_with_args
-
-                p = subprocess.Popen(cmd_line, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-                out, err = p.communicate()
-                self.valgrind_out = out.decode()
-
-    def is_passed(self):
-        if self.expected_output != 'ERROR':
-            with open(self.expected_output, 'r') as resultFile:
-                exp_out = resultFile.read()
-        else:
-            exp_out = 'ERROR\n'
-        return exp_out == self.actual_output and self.valgrind_out == ''
+        self.memcheck = memcheck
+        self.maxstack = maxstack
 
-    def get_log(self) -> str:
-        wrapped_args = ['"' + x + '"' for x in self.args[:-1]] + ["'" + self.args[-1] + "'"]
-        printable_args = ' '.join(wrapped_args)
+    def run_test(self):
 
-        if self.expected_output != 'ERROR':
-            with open(self.expected_output, 'r') as resultFile:
-                exp_out = resultFile.read()
-        else:
-            exp_out = 'ERROR\n'
+        # run the reference and user's test
+        ref_res = _run_test('sps-reference', self.args, self.file_input, False, False)
+        test_res = _run_test(self.exe_path, self.args, self.file_input, self.memcheck, self.maxstack)
 
-        passed = self.is_passed()
+        printable_args = ' '.join("'" + x + "'" if x != '-d' else x for x in self.args)
```

Comment:
Haven't seen this before, although it obviously works. Is that just a special list comprehension syntax I haven't seen before? Could you point me at where it is documented?

Reply:
It's just a python ternary statement inside a generator comprehension. Also could be written like this:
```py
printable_args = ' '.join([("'" + x + "'" if x != '-d' else x) for x in self.args])
```

Reply:
Ooh, generator comprehensions are a thing. The other bits were fine, it was just the lack of brackets that confused me :)